src/reflect/scala/reflect/runtime/JavaMirrors.scala

```diff
@@ -962,8 +962,9 @@ private[scala] trait JavaMirrors extends internal.SymbolTable with api.JavaUnive
       val enclosingClass = jclazz.getEnclosingClass
       var prefix = if (enclosingClass != null) enclosingClass.getName else ""
       val isObject = owner.isModuleClass && !owner.isPackageClass
-      if (isObject && !prefix.endsWith(nme.MODULE_SUFFIX_STRING)) prefix += nme.MODULE_SUFFIX_STRING
-      assert(jclazz.getName.startsWith(prefix))
+      val ownerIsJavaClass = owner.isJava && !owner.isPackageClass
+      if (isObject && !prefix.endsWith(nme.MODULE_SUFFIX_STRING) || ownerIsJavaClass) prefix += nme.MODULE_SUFFIX_STRING
+      assert(jclazz.getName.startsWith(prefix), (jclazz.getName, prefix))
       var name = jclazz.getName.substring(prefix.length)
       name = name.substring(name.lastIndexOf(".") + 1)
```

Comment:
This line is for non-nested classes, a comment would be useful. Or maybe better branch the method in the beginning with `if (enclosingClass != null) { .. } else { .. }`.